Add support for passing application information

ob-stripe opened the pull request:
r? @remi-stripe 
cc @stripe/api-libraries 

Adds a `StripeConfiguration.SetAppInfo()` method to provide application information that will be passed to Stripe in user-agent headers.

remi-stripe commented on src/Stripe.net/Infrastructure/Public/StripeConfiguration.cs:
```diff
@@ -176,6 +180,30 @@ public static JsonSerializerSettings DefaultSerializerSettings()
+        /// <param name="url">Website for your application (e.g. "https://myawesomeapp.info")</param>
+        public static void SetAppInfo(
+            string name,
+            string partnerId = null,
```
minor but other languages pass that one as the last one right?

remi-stripe commented on src/Stripe.net/Infrastructure/Public/SystemNetHttpClient.cs:
```diff
@@ -175,9 +152,38 @@ private static string BuildStripeClientUserAgentString()
+                {
+                    appInfo += $" ({StripeConfiguration.AppInfo["url"]})";
+                }
+
```
you forgot to handle partner id I think

remi-stripe commented on src/StripeTests/Infrastructure/Public/SystemNetHttpClientTest.cs:
```diff
@@ -31,5 +34,57 @@ public async Task MakeRequestAsync()
+
+            return userAgent.Contains("MyAwesomeApp/1.2.34 (https://myawesomeapp.info)") &&
+                appInfo.Value<string>("name") == "MyAwesomeApp" &&
+                appInfo.Value<string>("partner_id") == "pp_123" &&
```
Not sure how this test passed since you seem to not be setting this?

ob-stripe commented:
r? @remi-stripe 

I rewrote the implementation to use an actual class instead of a dictionary to store the values, and also added support for refreshing the user agent strings each time `SetAppInfo()` is called.

ob-stripe commented on src/Stripe.net/Infrastructure/Public/StripeConfiguration.cs:
```diff
@@ -176,6 +180,30 @@ public static JsonSerializerSettings DefaultSerializerSettings()
+                Url = url,
+            };
+
+            ((StripeClient as StripeClient)?.HttpClient as SystemNetHttpClient)?.RebuildUserAgentStrings();
```
This is pretty ugly, but I really don't want to change the `IStripeClient` and `IHttpClient` interfaces just to add support for this.

remi-stripe replied:
Yeah I don't understand that line. Mind asking someone else? :p

remi-stripe commented on src/Stripe.net/Infrastructure/Public/StripeConfiguration.cs:
```diff
@@ -176,6 +180,30 @@ public static JsonSerializerSettings DefaultSerializerSettings()
         }
 
+        /// <summary>
+        /// Sets information about your application. The information is passed along to Stripe.
```
So this is really about plugins/third-party libs here. But the wording of the description makes it look like any app should set this. Should we tweak it?
